src/Control/Category/Tensor.hs

```diff
@@ -1,220 +1,218 @@
 {-# LANGUAGE MonoLocalBinds #-}
 module Control.Category.Tensor where
 
-import Prelude hiding (id)
-import Control.Applicative
-import Control.Category (Category, id)
-import Data.Biapplicative
-import Data.Functor.Contravariant
-import Data.Profunctor
-import Data.These
-import Data.Void
+import           Control.Applicative
+import           Control.Category           (Category, id)
+import           Data.Biapplicative
+import           Data.Functor.Contravariant
+import           Data.Profunctor
+import           Data.These
+import           Data.Void
+import           Prelude                    hiding (id)
 
 {-
 
 | Tensor | Unit |
 +--------|------+
-| Either | Void |
 |  (,)   |  ()  |
+| Either | Void |
 | These  | Void |
 
-tensor = monoidal structure =
-    category
-  + bifunctor on that category `t`
-  + unit object in that category `i`
-  + isomorphisms in that category `t i a <-> a`, `t a i <-> a`, `t a (t b c) <-> t (t a b) c`
-  + some equalities
+tensor (or monoidal category) structure
+  = category `cat`
+  + bifunctor `t` on `cat`
+  + unit object `i` in `cat`
+  + isomorphisms in `cat`: `t i a <-> a`, `t a i <-> a`, `t a (t b c) <-> t (t a b) c`
+  + some equational laws
 
-monoidal functor = a functor that goes between the underlying
+lax monoidal functor = a functor that goes between the underlying
 categories of two different monoidal structure, and has a pair of
 operations `t2 (f a) (f b) -> f (t1 a b)` and `i2 -> f i1`
 
 -}
 
-class (Category cat1, Category cat2) => GBifunctor cat1 cat2 r t | t r -> cat1 cat2 where
-  gbimap :: a `cat1` b -> c `cat2` d -> t a c `r` t b d
+data Iso cat a b = Iso { fwd :: cat a b, bwd :: cat b a }
+
+-- Bifunctors
+class (Category cat1, Category cat2, Category cat3) => GBifunctor cat1 cat2 cat3 t | t cat3 -> cat1 cat2 where
+  gbimap :: cat1 a b -> cat2 c d -> cat3 (a `t` c)  (b `t` d)
+
+  infixr 9 #
+  (#) :: cat1 a b -> cat2 c d -> cat3 (a `t` c)  (b `t` d)
```

Comment:
Feels like this should live outside the class definition as a constrained top level definition, or we should have a `MINIMAL` annotation.

Reply:
how about the changes in the last commit?

Reply:
My personal preference is to move `(#)` outside the typeclass. Having it in the class opens up the possibility for an implementation that diverges from `gbimap`.

Reply:
what about for `\/` and `/\` ?

Reply:
Yeah its the same issue. It is nice to see them all together under the class, but it can lead to degenerate instances.